Add instructions to run coverage (#737)
ros2/ros2_documentation: source/Contributing/ROS-2-On-boarding-Guide.rst

  * In the first box "CI_BRANCH_TO_TEST" enter your feature branch name
  * Hit the ``build`` button

> * If your use case require to run code coverage:

Reviewer: ```suggestion
* If your use case requires running code coverage:
```

  * Go to ci.ros2.org
  * Log in (top right corner)
  * Click on the ``ci_linux_coverage`` job
  * Click "Build with Parameters" (left column)
  * In the "CI_ROS2_REPOS_URL" parameter use: `https://raw.githubusercontent.com/j-rivero/ros2/coverage/coverage.repos`
  * Leave "CI_BUILD_ARGS" and "CI_TEST_ARGS" with the default values
  * Hit the ``build`` button

>   * At the end of the document there are instructions about :ref:`how to interpret the result of repor <read-coverage-report>` and :ref:`calculate the coverage rate <calculate-coverage-rate>`

Reviewer 2: ```suggestion
  * At the end of the document there are instructions on how to :ref:`interpret the result of the report <read-coverage-report>` and :ref:`calculate the coverage rate <calculate-coverage-rate>`
```

      * results are exported as a cobertura report

* packaging (run every night, against fastrtps; result is bundled into an archive):

  * packaging_linux
  * packaging_osx
  * Packaging_windows

Note on Coverage runs
^^^^^^^^^^^^^^^^^^^^^

> ROS2 packages are organized in a way that the testing code for a given package is not only contained within the Package, but could also be present in a different package.

Reviewer 2: ```suggestion
ROS2 packages are organized in a way that the testing code for a given package is not only contained within the package, but could also be present in a different package.
```

> ROS2 packages are organized in a way that the testing code for a given package is not only contained within the package, but could also be present in a different package.

Reviewer 3: Use "ROS 2" - not "ROS2".

Author: https://github.com/ros2/ros2_documentation/pull/751

In other words: packages can exercise code belonging to other packages during the testing phase.

> To achieve the coverage rate reached by all code available in the ROS2 core packages it is recommended to run builds using a fixed set of proposed repositories.

Reviewer 2: ```suggestion
To achieve the coverage rate reached by all code available in the ROS 2 core packages it is recommended to run builds using a fixed set of proposed repositories.
```

That set is defined in the default parameters of coverage jobs in Jenkins.

.. _read-coverage-report:

How to read the coverage rate from the buildfarm report
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To see the coverage report for a given package:

> * When the ci_linux_coverage build finishes, click in ``Cobertura Coverage Report``

Reviewer: ```suggestion
* When the ``ci_linux_coverage`` build finishes, click in ``Coverage Report``
```

The button just says "Coverage Report" to me:

![covrep](https://user-images.githubusercontent.com/5751272/83823277-76f8c780-a688-11ea-8938-4c6919169e0f.png)

* Scroll down to the ``Coverage Breakdown by Package`` table
* In the table, look at the first column called "Name"

The coverage reports in the buildfarm include all the packages that were used in the ROS workspace.
The coverage report includes different paths corresponding to the same package:

* Name entries with the form: ``src.*.<repository_name>.<package_name>.*``

>   These corresponds to the unit test runs available in a package against its own source code

Reviewer 2: ```suggestion
  These correspond to the unit test runs available in a package against its own source code
```

* Name entries with the form: ``build.<repository_name>.<package_name>.*``
  These correspond to the unit test runs available in a package against its files generated at building or configuring time
* Name entries with the form: ``install.<package_name>.*``
  These correspond to the system/integration tests coming from testing runs of other packages

.. _calculate-coverage-rate:

How to calculate the coverage rate from the buildfarm report
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Get the combined unit coverage rate using the automatic script:

 * From the ci_linux_coverage Jenkins build copy the URL of the build
 * Download the `get_coverage_ros2_pkg <https://github.com/j-rivero/ros2_coverage_jenkins_params/blob/coverage_checker/get_coverage_results/get_coverage_ros2_pkg.py>`__ script

>  * Execute the script: ``./get_coverage_ros2_pkg.py <jenkins_build_url> <ros2_package_name>`` (`README <https://github.com/j-rivero/ros2_coverage_jenkins_params/blob/coverage_checker/get_coverage_results/README.md>`__)

Reviewer 3: Should we really recommend "external" scripts like this which have never been reviewed and are not accessible to edit by anyone else in the team?

Reviewer: @j-rivero is working on moving it to a more common location. Do you have a preference?

I reviewed the script while reviewing this PR.

Reviewer 3: > is working on moving it to a more common location. Do you have a preference?

> I reviewed the script while reviewing this PR.

Just a view things I notices while briefly looking at it (and since I don't see any PR on that repo I will just post it here):
* shebang line doesn't use `/usr/bin/env`
* incorrect usage handling
* error messages don't go to `stderr`
* 7 line convoluted logic to determine a simple boolean
* parentheses around conditional

Author: > I don't have a preference - just not on a personal GitHub user. Please suggest proposals.

Fully agree on moving it a better location. ros2/ci repository is the first one that comes to my mind.

> Just a view things I notices while briefly looking at it (and since I don't see any PR on that repo I will just post it here):

Thanks for looking into it Dirk. I've updated the code with your findings but let's start a PR procedure to look into it deeper.

 * Grab the results from the "Combined unit testing" final line in the output of the script

Alternative: get the combined unit coverage rate from coverage report (require manual calculation):

> * When the ci_linux_coverage build finishes, click in ``Cobertura Coverage Report``

Reviewer 2: ```suggestion
* When the ci_linux_coverage build finishes, click on ``Cobertura Coverage Report``
```

* Scroll down to the ``Coverage Breakdown by Package`` table
* In the table, under the first column "Name", look for (where <package_name> is your package under testing):

  * all the directories under the pattern ``src.*.<repository_name>.<package_name>.*`` grab the two absolute values in the column "Lines".
  * all the directories under the pattern ``build/.<repository_name>.*`` grab the two absolute values in the column "Lines".

* With the previous selection: for each cell, the first value is the lines tested and the second is the total lines of code.

>   Aggregate all rows for getting the total of the lines tested and the total of lines of code under test. Divide to get the coverage rate.

Reviewer 3: Wasn't this math incorrect (based on this comment: https://github.com/ros2/ros2_documentation/pull/737#discussion_r434532778)?

Author: We've taken the `install/` coverage out of the equation (by now) until we can merge properly the information coming from different types of testing but the results in `src.` and `build.` should be fine to aggregate.

Reviewer 3: > Aggregate all rows for getting the total of the lines tested and the total of lines of code under test. Divide to get the coverage rate.

But how is this calculation correct then? Couldn't it be that the same line is counted in both cases so shouldn't be summed up?

Author: > > Aggregate all rows for getting the total of the lines tested and the total of lines of code under test. Divide to get the coverage rate.
> 
> But how is this calculation correct then? Couldn't it be that the same line is counted in both cases so shouldn't be summed up?

With just that line of text the information could be misunderstood in the sense that "Aggregate all rows..." needs to be in the context of the sentence above (and previous):

>   * all the directories under the pattern ``src.*.<repository_name>.<package_name>.*`` grab the two absolute values in the column "Lines".
>  * all the directories under the pattern ``build/.<repository_name>.*`` grab the two absolute values in the column "Lines".
>  * With the previous selection: for each cell, the first value is the lines tested and the second is the total lines of code. 

As far I can tell, we are not duplicating the selection of coverage including `src/` and `build/`. Even in some cases where directory entries are named the same, files tested are different. Example: `rclcpp` with `node_interfaces` subdirectory present in both [src](https://ci.ros2.org/view/nightly/job/nightly_linux_coverage/884/cobertura/src_ros2_rclcpp_rclcpp_include_rclcpp_node_interfaces/) and [build]( https://ci.ros2.org/view/nightly/job/nightly_linux_coverage/884/cobertura/build_rclcpp_include_rclcpp_node_interfaces/ )

>   Aggregate all rows for getting the total of the lines tested and the total of lines of code under test. Divide to get the coverage rate.

Reviewer 3: Nitpick: new sentence - new line.

Author: https://github.com/ros2/ros2_documentation/pull/751

      * results are exported as a cobertura report

* packaging (run every night, against fastrtps; result is bundled into an archive):

  * packaging_linux
  * packaging_osx
  * Packaging_windows

Note on Coverage runs
^^^^^^^^^^^^^^^^^^^^^

ROS2 packages are organized in a way that the testing code for a given package
is not only contained within the Package, but could also be present in a

> different package. In other words: packages can exercice during the testing

Reviewer: ```suggestion
different package. In other words: packages can exercise during the testing
```

> phase code beloging to other packages.

Reviewer: ```suggestion
phase code belonging to other packages.
```

To achieve the coverage rate reached by all code available in the ROS2 core
packages it is recommended to run builds using a fixed set of proposed
repositories. That set is defined at https://raw.githubusercontent.com/j-rivero/ros2/coverage/coverage.repos

How to read the coverage rate from the buildfarm report
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To see the coverage report for a given package:

 * Go to the `Coverage Report` page as detailed in the Developer Workflow above
 * Scroll down to the `Coverage Breakdown by Package` table
 * In the table, look at the first column is called `Name`

> The coverage reports in the buildfarm include all the packages that were used in the ROS workspace. The coverage report include different paths corresponding to the same package:

Reviewer: ```suggestion
The coverage reports in the buildfarm include all the packages that were used in the ROS workspace. The coverage report includes different paths corresponding to the same package:
```

 1. Name entries with the form: `src.*.<repository_name>.<package_name>.*`
    These correspond the unit tests runs available in the own package against its own source code
 1. Name entries with the form: `build.<repository_name>.<package_name>.*`
    These correspond the unit tests runs available in the own package against its files generated at building or configuring time
 1. Name entries with the form: `install.<package_name>.*`
    These correspond to the system/integration tests coming from testing runs of other packages

> 

Reviewer: I think the only thing missing is how do we combine all these numbers to get to *the* code coverage for a package. I guess it probably `sum_of_lines_tested / sum_of_total_lines` from all the entries you mentioned?

Author: I've intentionally removed the sentence about that. The reason is simple, the solution I found is not. To properly combine unit, system/integration testing is not enough to just do `sum_of_lines_tested / sum_of_total_lines` . Imagining a 5 line program where unit and system testing (coming from other packages) are testing the same 1 line. If we add both we would get 2/5 while the reality is 1/5. To be able to merge that information in the right way I'm playing with gcovr merge features.

What I think we can add here is the fact that inside `build/` there are valid generated files that can be added that way since they should not be the same that the ones in `src/`.

      * results are exported as a cobertura report

* packaging (run every night, against fastrtps; result is bundled into an archive):

  * packaging_linux
  * packaging_osx
  * Packaging_windows

Note on Coverage runs
^^^^^^^^^^^^^^^^^^^^^

ROS2 packages are organized in a way that the testing code for a given package
is not only contained within the Package, but could also be present in a
different package. In other words: packages can exercise during the testing

> phase code belonging to other packages.

Reviewer 2: ```suggestion
different package. In other words: packages can exercise code belonging to other packages during the testing phase.
```

To achieve the coverage rate reached by all code available in the ROS2 core
packages it is recommended to run builds using a fixed set of proposed
repositories. That set is defined at https://raw.githubusercontent.com/j-rivero/ros2/coverage/coverage.repos

.. _read-coverage-report:

How to read the coverage rate from the buildfarm report
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To see the coverage report for a given package:

* When the ci_linux_coverage build finishes, click in ``Cobertura Coverage Report``
* Scroll down to the ``Coverage Breakdown by Package`` table

> * In the table, look at the first column is called "Name"

Reviewer 2: ```suggestion
* In the table, look at the first column called "Name"
```

> The coverage reports in the buildfarm include all the packages that were used in the ROS workspace. The coverage report includes different paths corresponding to the same package:

Reviewer 3: Nitpick: follow style guide for one sentence per line.

Author: thanks, 1fc2b56

* Name entries with the form: ``src.*.<repository_name>.<package_name>.*``

>   These correspond the unit tests runs available in the own package against its own source code

Reviewer 2: ```suggestion
  These corresponds to the unit test runs available in a package against its own source code
```

* Name entries with the form: ``build.<repository_name>.<package_name>.*``

>   These correspond the unit tests runs available in the own package against its files generated at building or configuring time

Reviewer 2: ```suggestion
  These correspond to the unit test runs available in a package against its files generated at building or configuring time
```

* Name entries with the form: ``install.<package_name>.*``
  These correspond to the system/integration tests coming from testing runs of other packages

.. _calculate-coverage-rate:

How to calculate the coverage rate from the buildfarm report
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To get the combined unit coverage rate:

* When the ci_linux_coverage build finishes, click in ``Cobertura Coverage Report``
* Scroll down to the ``Coverage Breakdown by Package`` table

> * In the table, the first column is colled "Name", look for (where <package_name> is your package under testing):

Reviewer 2: ```suggestion
* In the table, under the first column "Name", look for (where <package_name> is your package under testing):
```
